src/script/sign.h

```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+                    }
+                    s >> redeem_script;
+                    break;
+                }
+                case PSBT_IN_WITNESSSCRIPT:
+                {
+                    if (!witness_script.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, input witnessScript already provided");
+                    }
+                    s >> witness_script;
+                    break;
+                }
+                case PSBT_IN_BIP32_DERIVATION:
+                {
+                    // Make sure that the key is the size of pubkey + 1
+                    if (key.size() != CPubKey::PUBLIC_KEY_SIZE + 1 && key.size() != CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1) {
+                        throw std::ios_base::failure("Size of key was not the expected size for the type BIP32 keypath");
+                    }
+                    // Read in the pubkey from key
+                    CPubKey pubkey(key.begin() + 1, key.end());
+                    if (!pubkey.IsFullyValid()) {
+                       throw std::ios_base::failure("Invalid pubkey");
+                    }
+                    if (hd_keypaths.count(pubkey) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, input pubkey derivation path already provided");
+                    }
+
+                    // Read in key path
+                    uint64_t value_len = ReadCompactSize(s);
+                    std::vector<uint32_t> keypath;
+                    for (unsigned int i = 0; i < value_len; i += sizeof(uint32_t)) {
+                        uint32_t index;
+                        s >> index;
+                        keypath.push_back(index);
+                    }
+
+                    // Add to map
+                    hd_keypaths.emplace(pubkey, keypath);
+                    break;
+                }
+                case PSBT_IN_SCRIPTSIG:
+                {
+                    if (!final_script_sig.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, input final scriptSig already provided");
+                    }
+                    s >> final_script_sig;
+                    break;
+                }
+                case PSBT_IN_SCRIPTWITNESS:
+                {
+                    if (!final_script_witness.IsNull()) {
+                        throw std::ios_base::failure("Duplicate Key, input final scriptWitness already provided");
+                    }
+                    UnserializeFromVector(s, final_script_witness.stack);
+                    break;
+                }
+                // Unknown stuff
+                default:
+                    if (unknown.count(key) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, key for unknown value already provided");
+                    }
+                    // Read in the value
+                    uint64_t value_len = ReadCompactSize(s);
```

Comment:
Here you can again just use `std::vector<unsigned char> val_bytes; s >> val_bytes;`.

Reply:
Done